python/cpython
# gh-104090: Update Resource Tracker to return exit code based on resource leaked found or not

1. Add tests to catch the resource leak bug in `collectedDurations`  and verified that it fails without the fix
<details>
<summary>Show output</summary>

```shell
$ ./python -m unittest test.test_concurrent_futures.FailingInitializerResourcesTest
.
/home/user/cpython/Lib/multiprocessing/resource_tracker.py:228: UserWarning: resource_tracker: There appear to be 3 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
FTraceback (most recent call last):
  File "/home/user/cpython/Lib/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()
  File "/home/user/cpython/Lib/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/cpython/Lib/multiprocessing/synchronize.py", line 87, in _cleanup
    sem_unlink(name)
FileNotFoundError: [Errno 2] No such file or directory
.
.
.
======================================================================
FAIL: test_forkserver (test.test_concurrent_futures.FailingInitializerResourcesTest.test_forkserver)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/cpython/Lib/test/test_concurrent_futures.py", line 318, in test_forkserver
    self._test(ProcessPoolForkserverFailingInitializerTest)
  File "/home/user/cpython/Lib/test/test_concurrent_futures.py", line 312, in _test
    self.assertEqual(_resource_tracker._exitcode, 0)
AssertionError: 256 != 0

======================================================================
FAIL: test_spawn (test.test_concurrent_futures.FailingInitializerResourcesTest.test_spawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/user/cpython/Lib/test/test_concurrent_futures.py", line 315, in test_spawn
    self._test(ProcessPoolSpawnFailingInitializerTest)
  File "/home/user/cpython/Lib/test/test_concurrent_futures.py", line 312, in _test
    self.assertEqual(_resource_tracker._exitcode, 0)
AssertionError: 256 != 0

----------------------------------------------------------------------
Ran 2 tests in 0.727s

FAILED (failures=2)
0.33s
```
</details>


2. Add returning exit code from `resource_tracker.py` based on the cleanup result, so we can check what was the result
3. Add tests for Resource Tracker exit code based on leaked/no leaked resources



* Issue: gh-104090
